addons/survey/controllers/main.py

```diff
@@ -122,28 +124,38 @@ def _redirect_with_error(self, access_data, error_key):
         elif error_key == 'survey_closed' and access_data['can_answer']:
             return request.render("survey.survey_expired", {'survey': survey_sudo})
         elif error_key == 'survey_auth' and answer_sudo.token:
-            return request.render("survey.auth_required", {'survey': survey_sudo, 'token': answer_sudo.token})
+            if answer_sudo.partner_id and (answer_sudo.partner_id.user_ids or survey_sudo.users_can_signup):
+                if answer_sudo.partner_id.user_ids:
+                    answer_sudo.partner_id.signup_cancel()
+                else:
+                    answer_sudo.partner_id.signup_prepare(expiration=fields.Datetime.now() + relativedelta(days=1))
+                redirect_url = answer_sudo.partner_id._get_signup_url_for_action(url='/survey/start/%s?token=%s' % (survey_sudo.access_token, answer_sudo.token))[answer_sudo.partner_id.id]
+            else:
+                redirect_url = '/web/login?redirect=%s' % ('/survey/start/%s?token=%s' % (survey_sudo.access_token, answer_sudo.token))
+            return request.render("survey.auth_required", {'survey': survey_sudo, 'redirect_url': redirect_url})
         elif error_key == 'answer_deadline' and answer_sudo.token:
             return request.render("survey.survey_expired", {'survey': survey_sudo})
+        elif error_key == 'answer_done' and answer_sudo.token:
+            return request.render("survey.sfinished", {'survey': survey_sudo, 'token': answer_sudo.token, 'user_input': answer_sudo})
 
         return werkzeug.utils.redirect("/")
 
-    @http.route('/survey/test/<int:survey_id>', type='http', auth='user', website=True)
-    def survey_test(self, survey_id, token=None, **kwargs):
+    @http.route('/survey/test/<string:survey_token>', type='http', auth='user', website=True)
+    def survey_test(self, survey_token, **kwargs):
         """ Test mode for surveys: create a test answer, only for managers or officers
         testing their surveys """
-        survey_sudo = request.env['survey.survey'].sudo().browse(survey_id)
+        survey_sudo, dummy = self._fetch_from_access_token(survey_token, False)
         try:
             answer_sudo = survey_sudo._create_answer(user=request.env.user, test_entry=True)
         except:
             return werkzeug.utils.redirect('/')
-        return request.redirect('/survey/start/%s?%s' % (survey_sudo.id, keep_query('*', token=answer_sudo.token)))
+        return request.redirect('/survey/start/%s?%s' % (survey_sudo.access_token, keep_query('*', token=answer_sudo.token)))
 
-    @http.route('/survey/start/<int:survey_id>', type='http', auth='public', website=True)
-    def survey_start(self, survey_id, token=None, email=False, **post):
+    @http.route('/survey/start/<string:survey_token>', type='http', auth='public', website=True)
+    def survey_start(self, survey_token, token=None, email=False, **post):
```

Comment:
@tde-banana-odoo Maybe rename the second token parameter "answer_token" to make it less confusing? 
If so it needs to be done in several controller methods.

Reply:
I had the same thought and did not do it because I am a slacker. Seems it did not pay off.

Reply:
Done.